# Source of Error 

In this lesson, we'll begin practice some of the techniques for building a simulation dataset.  

The NYTimes is considering hiring us to predict the popularity of their articles.  We'd like to give them a sense of the predictions that we could make even before we get our hands on their data.  To do so we construct a sample dataset.

### Constructing the dataset

Declare a method called `build_feature` that returns a feature vector.  The method takes an argument of `size`, which specifies the length, or dimension, of the vector returned.  It also has a second argument of `going_from` which specifies the range of the vector.

In [1]:
import numpy as np 
# use numpy's randint feature

def build_feature(size, going_from = []):
    return np.random.randint(going_from[0], going_from[1], size=size) 

In [2]:
np.random.seed(1)
word_counts = build_feature(50, [300, 1000])
word_counts

array([337, 535, 372, 945, 444, 429, 883, 808, 690, 581, 478, 576, 554,
       657, 768, 552, 790, 968, 698, 862, 880, 515, 803, 778, 386, 441,
       693, 307, 619, 834, 613, 813, 616, 509, 564, 953, 927, 731, 933,
       756, 842, 371, 687, 754, 861, 613, 815, 797, 343, 888])

In [3]:
len(word_counts)
# 50

50

In [4]:
word_counts = word_counts.reshape(-1, 1)
word_counts[0:5]

array([[337],
       [535],
       [372],
       [945],
       [444]])

In [5]:
word_counts.shape
# (50, 1)
# The feature word_counts should return a vector of rows

(50, 1)

In [6]:
word_counts[0:3]
# array([[337],
#        [535],
#        [372]])

array([[337],
       [535],
       [372]])

### Building the targets variables

Now, imagine that we are told that the true model, which determines the number of clicks on a news article, is the following: 

$y = \theta_1x_1 + \theta_0$

where:

* $y$ is a vector that contains the actual clicks on each article
* $x_1$ is a vector of word counts of each article and 
* $\theta_0$ represents the baseline number of clicks that an article receives (regardless of the word count)

We are also provided the following:

* $\theta_1 = 75$
* $\theta_0 = 10000$

Using the `word_counts` vector that we constructed above, and the defined parameter values, create our vector of target variables given the true model defined above. 

In [7]:
# Hint: break the problem into steps, examine the hstack method

In [8]:
word_counts[0:5]

array([[337],
       [535],
       [372],
       [945],
       [444]])

In [9]:
biases = np.ones(50).reshape(-1, 1)
biases[0:5]

array([[1.],
       [1.],
       [1.],
       [1.],
       [1.]])

In [10]:
features = np.hstack((word_counts, biases))
features[0:5]

array([[337.,   1.],
       [535.,   1.],
       [372.,   1.],
       [945.,   1.],
       [444.,   1.]])

In [11]:
features.shape

(50, 2)

In [12]:
params = np.array([75, 10000])
params

array([   75, 10000])

In [13]:
params.shape

(2,)

In [14]:
(features @ params).shape

(50,)

In [15]:
clicks = features @ params

In [16]:
clicks
# array([35275., 50125., 37900., 80875., 43300., 42175., 76225., 70600.,
#        61750., 53575., 45850., 53200., 51550., 59275., 67600., 51400.,
#        69250., 82600., 62350., 74650., 76000., 48625., 70225., 68350.,
#        38950., 43075., 61975., 33025., 56425., 72550.])

array([35275., 50125., 37900., 80875., 43300., 42175., 76225., 70600.,
       61750., 53575., 45850., 53200., 51550., 59275., 67600., 51400.,
       69250., 82600., 62350., 74650., 76000., 48625., 70225., 68350.,
       38950., 43075., 61975., 33025., 56425., 72550., 55975., 70975.,
       56200., 48175., 52300., 81475., 79525., 64825., 79975., 66700.,
       73150., 37825., 61525., 66550., 74575., 55975., 71125., 69775.,
       35725., 76600.])

### Training our model

Now that we have our dataset, let's train our model on the constructed dataset.  Assign the model to the variable `model`.

In [17]:
# remember to use set fit_intercept=False

from sklearn.linear_model import LinearRegression

model = LinearRegression(fit_intercept=False)

model.fit(features, clicks)

LinearRegression(copy_X=True, fit_intercept=False, n_jobs=None,
         normalize=False)

In [18]:
model.coef_
# array([   75., 10000.])

array([   75., 10000.])

### Exploring irreducible error

At this point we now have trained a model to discover the true feature parameters of the underlying model.  Let's now consider, more realistically, that the underlying model takes the following form.

$y = \theta_1x_1 + \theta_0 + \epsilon_i$

Where 
* $\theta_1 = 75$
* $\theta_0 = 10000$

Use numpy's `randint` function to create a vector of random errors that range between positive and negative 8000.  Use this to create a new vector of target variables called `noisy_clicks` of length 50.

In [19]:
np.random.seed(1)
random_errors = np.random.randint(-8000, 8000, 50)
random_errors

array([ 5349, -7765,  4172, -2808, -7095,  2955,  -187, -5105, -2944,
       -7856, -3775,  -249,  2989,  6844, -4538,  7641,  1394, -2604,
       -2626,  4645, -5038, -5484,  7243,   444, -4438, -3236,    93,
       -1458, -7438,  2820,  7575,   151, -4951, -7247,  1719,  3742,
       -6112, -6890, -1715, -1879,  7753, -6969, -3585, -5123, -4394,
        1529, -5439,  6208,  4604,  6545])

In [20]:
noisy_clicks = clicks + random_errors
noisy_clicks

array([40624., 42360., 42072., 78067., 36205., 45130., 76038., 65495.,
       58806., 45719., 42075., 52951., 54539., 66119., 63062., 59041.,
       70644., 79996., 59724., 79295., 70962., 43141., 77468., 68794.,
       34512., 39839., 62068., 31567., 48987., 75370., 63550., 71126.,
       51249., 40928., 54019., 85217., 73413., 57935., 78260., 64821.,
       80903., 30856., 57940., 61427., 70181., 57504., 65686., 75983.,
       40329., 83145.])

In [21]:
noisy_clicks[0:10]
# array([40624., 42360., 42072., 78067., 36205., 45130., 76038., 65495.,
#        58806., 45719.])

array([40624., 42360., 42072., 78067., 36205., 45130., 76038., 65495.,
       58806., 45719.])

In [22]:
# if you have the plotly jupyter lab extension loaded, try the following
from graph import trace_values, plot
noisy_customers_trace = trace_values(word_counts.reshape(50), noisy_clicks)
plot([noisy_customers_trace])

Now we have a model that has discovered the parameters of our model:

$y = \theta_1x_1 + \theta_0 + \epsilon_i$

Where 
* $\theta_1 = 75$
* $\theta_0 = 10000$

Still, that $\epsilon_i$ represents our irreducible error, that is inherent to the future outcomes we try to predict.

Let's see this.  Once again, these are the parameters of our model.

In [23]:
model.coef_
# array([   75., 10000.])

array([   75., 10000.])

Use these parameters along with our features to predict our data.  Use the sklearn's `predict` function in the linear regression class to make the predictions. 

In [24]:
predictions = model.predict(features)
predictions
# array([35275., 50125., 37900., 80875., 43300., 42175., 76225., 70600.,
#        61750., 53575., 45850., 53200., 51550., 59275., 67600., 51400.,
#        69250., 82600., 62350., 74650., 76000., 48625., 70225., 68350.,
#        38950., 43075., 61975., 33025., 56425., 72550., 55975., 70975.,
#        56200., 48175., 52300., 81475., 79525., 64825., 79975., 66700.,
#        73150., 37825., 61525., 66550., 74575., 55975., 71125., 69775.,
#        35725., 76600.])

array([35275., 50125., 37900., 80875., 43300., 42175., 76225., 70600.,
       61750., 53575., 45850., 53200., 51550., 59275., 67600., 51400.,
       69250., 82600., 62350., 74650., 76000., 48625., 70225., 68350.,
       38950., 43075., 61975., 33025., 56425., 72550., 55975., 70975.,
       56200., 48175., 52300., 81475., 79525., 64825., 79975., 66700.,
       73150., 37825., 61525., 66550., 74575., 55975., 71125., 69775.,
       35725., 76600.])

> Show that these predictions are the same that we get from matrix vector multiplication: $h(x) = X \cdot \theta$

In [25]:
matrix_vector_predictions = features @ params
matrix_vector_predictions
# array([35275., 50125., 37900., 80875., 43300., 42175., 76225., 70600.,
#        61750., 53575., 45850., 53200., 51550., 59275., 67600., 51400.,
#        69250., 82600., 62350., 74650., 76000., 48625., 70225., 68350.,
#        38950., 43075., 61975., 33025., 56425., 72550., 55975., 70975.,
#        56200., 48175., 52300., 81475., 79525., 64825., 79975., 66700.,
#        73150., 37825., 61525., 66550., 74575., 55975., 71125., 69775.,
#        35725., 76600.])

array([35275., 50125., 37900., 80875., 43300., 42175., 76225., 70600.,
       61750., 53575., 45850., 53200., 51550., 59275., 67600., 51400.,
       69250., 82600., 62350., 74650., 76000., 48625., 70225., 68350.,
       38950., 43075., 61975., 33025., 56425., 72550., 55975., 70975.,
       56200., 48175., 52300., 81475., 79525., 64825., 79975., 66700.,
       73150., 37825., 61525., 66550., 74575., 55975., 71125., 69775.,
       35725., 76600.])

Ok, now let's use linear algebra to calculate the RSS of our model applied to predicting our noisy_clicks.

In [26]:
actual = clicks + random_errors
expected = predictions

actual_minus_expected = actual - expected
actual_minus_expected
# array([ 5349., -7765.,  4172., -2808., -7095.,  2955.,  -187., -5105.,
#        -2944., -7856., -3775.,  -249.,  2989.,  6844., -4538.,  7641.,
#         1394., -2604., -2626.,  4645., -5038., -5484.,  7243.,   444.,
#        -4438., -3236.,    93., -1458., -7438.,  2820.,  7575.,   151.,
#        -4951., -7247.,  1719.,  3742., -6112., -6890., -1715., -1879.,
#         7753., -6969., -3585., -5123., -4394.,  1529., -5439.,  6208.,
#         4604.,  6545.])

array([ 5349., -7765.,  4172., -2808., -7095.,  2955.,  -187., -5105.,
       -2944., -7856., -3775.,  -249.,  2989.,  6844., -4538.,  7641.,
        1394., -2604., -2626.,  4645., -5038., -5484.,  7243.,   444.,
       -4438., -3236.,    93., -1458., -7438.,  2820.,  7575.,   151.,
       -4951., -7247.,  1719.,  3742., -6112., -6890., -1715., -1879.,
        7753., -6969., -3585., -5123., -4394.,  1529., -5439.,  6208.,
        4604.,  6545.])

Now calculate rss using matrix algebra.

In [27]:
rss = actual_minus_expected @ actual_minus_expected
rss
# 1203968931.0

1203968931.0

From here we can calculate or mean squared error, simply by dividing the residual sum of squares by the number of predictions.

In [28]:
mse = (rss/len(predictions))
mse
# 24079378.62

24079378.62

Of course, we could have just used the sklearn library to calculate this for us.

In [29]:
from sklearn.metrics import mean_squared_error
mean_squared_error(noisy_clicks, predictions)
# 24079378.619999994

24079378.619999994

Now if you remember RSS, RSS squares the error and then adds up these errors.  So, with mean squared error, we now just take the average of RSS.  But it's still hard to interpret squared error.  A more understandable metric is the root mean squared error, which is just the square root of the average squared error.  We'll calculate it below. 

In [30]:
from math import sqrt
sqrt(mean_squared_error(predictions, noisy_clicks))

4907.074344250349

So this says that on average, we are off by 4907 when we try to predict the noisy clicks.  This makes sense, as this is close to the average of our irreducible error.

In [31]:
np.average(abs(random_errors))
#4307.26

4307.26

## Working with Variance

Now produce a model that not only suffers from irreducible error, but also suffers from variance.  Use the `noisy_clicks` target variables and the previously defined `features` to produce this model.  Assign it to the variable `model_with_variance`.


In [32]:
model_with_variance = LinearRegression(fit_intercept = False).fit(features, noisy_clicks)

In [33]:
model_with_variance

LinearRegression(copy_X=True, fit_intercept=False, n_jobs=None,
         normalize=False)

In [34]:
model_with_variance.coef_
# array([  76.53450394, 8119.10474367])

array([  76.53450394, 8119.10474367])

So comparing with our first model, we see that our parameters no longer match the underlying values.

In [35]:
model.coef_

array([   75., 10000.])

### Predicting with our models

We now have two models - one with irreducible error, and another that suffers from irreducible error and variance.  Let's compare these models.

We already calculated the root mean squared error for our first model.  We saw that it fits the training set perfectly (or close to it).

In [36]:
from math import sqrt
sqrt(mean_squared_error(predictions, clicks))
#9.203439287865707e-12

9.203439287865707e-12

Now let's generate a new vector of target variables to see how our original model and our model with variance performs.  To do so, we'll first generate a new set of errors.

In [37]:
np.random.seed(10)
new_errors = np.random.randint(-8000, 8000, 50)
new_errors

array([-6711,  -707,  4815, -6656,  3633,  -709,  1372,  2201, -3171,
       -6480,  2141,  2234,  1224,  1289, -1600,  6826,  4328,  2742,
        7780,  6707, -2352, -3548, -7761, -5557, -5898, -4584,  3627,
        -710, -7426,  7905,  5512,    36,  1166,  6001,  3318,  4365,
        7813,  4301, -1101, -7591, -6594, -7347, -4068, -6634,  4062,
        7006,   281, -5300,  -511,  6879])

Now use the new errors to generate a new vector of observed data.  Data that neither model been trained with. 

In [38]:
new_noisy_clicks = clicks + new_errors
new_noisy_clicks

array([28564., 49418., 42715., 74219., 46933., 41466., 77597., 72801.,
       58579., 47095., 47991., 55434., 52774., 60564., 66000., 58226.,
       73578., 85342., 70130., 81357., 73648., 45077., 62464., 62793.,
       33052., 38491., 65602., 32315., 48999., 80455., 61487., 71011.,
       57366., 54176., 55618., 85840., 87338., 69126., 78874., 59109.,
       66556., 30478., 57457., 59916., 78637., 62981., 71406., 64475.,
       35214., 83479.])

Now calculate the root mean squared error for both the original model and the model with variance on this dataset.

In [39]:
rmse_new_noisy_clicks_original_model = sqrt(mean_squared_error(new_noisy_clicks, predictions))
rmse_new_noisy_clicks_original_model
# 4921.427288907152

4921.427288907152

In [40]:
variance_predictions = model_with_variance.predict(features)
variance_predictions

array([33911.232572  , 49065.06435244, 36589.94020995, 80444.21096851,
       42100.42449375, 40952.40693463, 75699.07172413, 69958.98392851,
       60927.91246339, 52585.65153376, 44702.59762777, 52202.97901405,
       50519.21992733, 58402.27383332, 66897.60377084, 50366.15091945,
       68581.36285756, 82204.50455917, 61540.18849493, 74091.84714135,
       75469.4682123 , 47534.37427361, 69576.3114088 , 67662.94881026,
       37661.42326514, 41870.82098193, 61157.51597522, 31615.19745375,
       55493.96268354, 71948.88103099, 55034.75565989, 70341.65644821,
       55264.35917171, 47075.16724996, 51284.56496675, 81056.48700004,
       79066.58989756, 64065.827125  , 79525.79692121, 65979.18972354,
       72561.15706252, 36513.40570601, 60698.30895157, 65826.12071566,
       74015.31263741, 55034.75565989, 70494.7254561 , 69117.10438515,
       34370.43959565, 76081.74424384])

In [41]:
rmse_new_noisy_clicks_variance_model = sqrt(mean_squared_error(new_noisy_clicks, variance_predictions))
rmse_new_noisy_clicks_variance_model
# 4984.958232759586

4984.958232759586

Notice that our model did worse on this new dataset than it did with the data that it trained on.

In [42]:
from math import sqrt
sqrt(mean_squared_error(noisy_clicks, variance_predictions))
#4824.297529089475

4824.297529089475

This is generally the case.  When we train the data, our linear regression model is trying to fit to the training data.  And thus it finds parameters that match some of the randomness in the training data.  However, because this randomness is not replicated in future data, it generally performs less well on data it has not seen.  When seeing how well our model will perform, we want to see how well it performs on data it has not trained on, as this better simulates how the model will perform in production.

### Summary

Great job!  We used this lesson to review our teachings in linear algebra and numpy.  We also used it to develop our understanding of types of error: both variance and irreducible error.  We saw that if we only assess how well a model performs on the training set, this performance likely will not generalize to future data as our model has likely overfit to the training set.  To remedy this, we should assess how well a model performs on data it has not trained on.